File: hospital_occupancy_tracker.R
########################################################################
# hospital occupancy tracker
# reads file from web, sorts/processes data, saves new data to csv, plots data

# load packages
library(tidyverse)

url <- "https://www.msss.gouv.qc.ca/professionnels/statistiques/documents/urgences/Releve_horaire_urgences_7jours.csv"

# read csv file with french characters
df <- read.csv(url, encoding = "latin1")

update <- as.Date(df$Mise_a_jour[1])
update_time <- df$Heure_de_l.extraction_.image.[1]

# select montreal hospitals
df <- df %>% select(etab = Nom_etablissement, hospital_name = Nom_installation, beds_total = Nombre_de_civieres_fonctionnelles, beds_occ = Nombre_de_civieres_occupees) %>%
  filter(str_detect(etab, "Montr|CHUM|CUSM|CHU Sainte-Justine")) %>%
  mutate(beds_total = as.numeric(beds_total), beds_occ = as.numeric(beds_occ)) %>% 
  select(hospital_name, beds_occ, beds_total)
# calculate total and add to df 
df %>% summarise(sum(beds_total), sum(beds_occ)) -> total
df <- df %>% add_row(hospital_name = "Total", beds_occ = total[1,2], beds_total = total[1,1] ) %>%
  mutate(occupancy_rate = round(100*(beds_occ/beds_total)), Date = update) %>%
  select(Date, hospital_name, occupancy_rate)

# write row to csv
row <- df %>% pivot_wider(names_from=hospital_name, values_from=occupancy_rate)
# write.csv(row, file = "data/hospitals.csv", row.names = FALSE) # creates file, run only once!
write.table(row, "data/hospitals.csv", append = T, row.names = F, col.names = F, sep = ",")

# visualization

update_txt <- paste("\nlast update:", update, "at", update_time)
df %>% 
  filter(hospital_name != "Total") %>%
  ggplot(aes(x = reorder(hospital_name, occupancy_rate), y = occupancy_rate, fill = occupancy_rate)) + 
  geom_col(position = "identity", size = 0.5, show.legend = F) +
  geom_text(aes(label = paste0(occupancy_rate,"%")), hjust = 1, colour = "white", size = 3) +
  coord_flip() +
  scale_fill_gradient2(low = "light green", high = "red", mid = "yellow", midpoint = 80) + 
  theme_minimal() +
  labs(caption = update_txt, x = NULL, y = NULL) +
  theme(panel.grid.minor = element_blank())

ggsave("img/today.png")
ggsave("img/today.jpeg")

# don't create pdf
pdf(NULL)

# to do:
# visualization tracker 

File: hospital_occupancy_tracker.py
import pandas as pd
import numpy as np

# read data
url = "https://www.msss.gouv.qc.ca/professionnels/statistiques/documents/urgences/Releve_horaire_urgences_7jours.csv"
df = pd.read_csv(url, encoding="latin1")

# get current date and time
update = df.iloc[0, 8].split("T")[0]
update_time = df.iloc[0, 8].split("T")[1]

# filter (in R: select) columns
# obs! column-names contain whitespaces
df = df.filter(items=['Nom_etablissement ', ' Nom_installation ', ' Nombre_de_civieres_fonctionnelles ', ' Nombre_de_civieres_occupees '])

# get montreal hospitals (in R filter)
df = df.loc[df['Nom_etablissement '].str.contains(r"Montr|CHUM|CUSM|CHU Sainte-Justine")]

# replace missings with NaNs
df[' Nombre_de_civieres_fonctionnelles '] = pd.to_numeric(df[' Nombre_de_civieres_fonctionnelles '], errors='coerce')
df[' Nombre_de_civieres_occupees '] = pd.to_numeric(df[' Nombre_de_civieres_occupees '], errors='coerce')

# get stretchers/civieres data
beds_total = df[' Nombre_de_civieres_fonctionnelles '].sum()
beds_occ = df[' Nombre_de_civieres_occupees '].sum()

# add new row to end of dataframe, containing beds_total and beds_occ
df.loc[df.index[-1]+1, ['Nom_etablissement ', ' Nom_installation ', ' Nombre_de_civieres_fonctionnelles ', ' Nombre_de_civieres_occupees ']] = ['Total', 'Total', beds_total, beds_occ]

# calculate occupancy rate if not NaN
f = lambda x: round(100*(x[' Nombre_de_civieres_occupees ']/x[' Nombre_de_civieres_fonctionnelles '])) if x.notnull().all() else np.NaN

# add new column with occupancy_rate
df['occupancy_rate'] = df.apply(f, axis=1)

# calculate occupancy_rate
# df = df.assign(occupancy_rate = lambda x: round(100*(x[' Nombre_de_civieres_occupees ']/x[' Nombre_de_civieres_fonctionnelles '])))

# transform occupancy_rate to integer
# df["occupancy_rate"] = df["occupancy_rate"].astype(int)
# OBS! in python NaN is a float > convert to Int64
df['occupancy_rate'] = pd.to_numeric(df['occupancy_rate'], downcast='integer', errors='coerce').astype('Int64')

# insert new column containing the current date
df = df.assign(Date=update)

# select three columns
df = df.filter(items=['Date', ' Nom_installation ', 'occupancy_rate'])

# rename column to "hospital_name"
df.rename(columns={' Nom_installation ': 'hospital_name'}, inplace=True)

# transform from long to wide format (pivot)
df = pd.pivot(df, index='Date', columns='hospital_name', values='occupancy_rate')

# after transforming, python reorders columns in alphabetically order whereas R leaves them in the same order as the rows were
# reorder columns, so they will be in the same order as in R
df = df.reindex(columns=["Centre hospitalier de St. Mary", "Institut universitaire en santé mentale Douglas", "Hôpital de LaSalle", "Hôpital général du Lakeshore", "L'Hôpital général Juif Sir Mortimer B. Davis", "Hôpital de Verdun", "Hôpital Notre-Dame", "Hôpital Fleury", "Hôpital du Sacré-Coeur de Montréal", "Hôpital Jean-Talon", "Pavillon Albert-Prévost", "Institut universitaire en santé mentale de Montréal", "Hôpital Maisonneuve-Rosemont", "Hôpital Santa Cabrini", "Centre hospitalier de l'Université de Montréal", "Hôpital Royal Victoria", "Hôpital général de Montréal", "Hôpital de Montréal pour enfants", "Campus Lachine", "Institut de Cardiologie de Montréal", "CHU Sainte-Justine", "Total"])

# write row to csv (a = append), replace missings with "NA"
df.to_csv("data/hospitals.csv", header=False, mode="a",  na_rep='NA')

# Voilà! exaclty the same output as in R
